src/boot/efi/boot.c

```diff
@@ -2013,6 +2074,80 @@ static VOID config_entry_add_linux(
         uefi_call_wrapper(linux_dir->Close, 1, linux_dir);
 }
 
+static EFI_STATUS secure_boot_enroll(Config *config, ConfigEntry *entry) {
+        EFI_STATUS err = EFI_SUCCESS;
+        static const UINT32 sb_vars_opts =
+                EFI_VARIABLE_NON_VOLATILE |
+                EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                EFI_VARIABLE_RUNTIME_ACCESS |
+                EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+                0;
+        SecureBootVariable *vars = entry->sb_vars;
+
+        /* we then write them to the efivars */
+        for (UINTN i = 0; i < ELEMENTSOF(sb_vars); i++) {
+                err = efivar_set_raw(&vars[i].vendor, vars[i].name, vars[i].buffer, vars[i].size, sb_vars_opts);
+                if (EFI_ERROR(err)) {
+                        Print(L"Failed to write %s secure boot variable.\n", vars[i].name);
+                        return err;
+                }
+                FreePool(vars[i].buffer);
+        }
+
+        /* if we loaded the signature keys successfully then we remove all the
+         * signature keys entries from  the menu as the system is now locked down. */
+        config_remove_by_type(config, LOADER_SIGNATURE_KEYS);
+
+        return err;
+}
+
+static ConfigEntry *config_entry_add_signature_keys(
+                EFI_FILE *root_dir,
+                Config *config,
+                CHAR16 *name) {
+        EFI_STATUS err = EFI_SUCCESS;
+        SecureBootVariable *vars = NULL;
+        ConfigEntry *entry = NULL;
+        CHAR16 *title = NULL, *id = NULL, *keys_dir = NULL;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE dir = NULL;
+
+        keys_dir = PoolPrint(L"\\loader\\keys\\%s", name);
+        err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &dir, keys_dir, EFI_FILE_MODE_READ, 0ULL);
+        if (EFI_ERROR(err)) {
+                return NULL;
+        }
+
+        vars = AllocatePool(sizeof(sb_vars));
```

Comment:
I know we're not checking every allocation right now, but I intend to fix that at some point. For now we should make sure to check allocations are successful.

Reply:
yes, we should check oom on EFI too. some parts of the codebase already does, but most does not. And it should do so everywhere, to match the userspace codebase